contracts/meta/MetaSwap.sol

```diff
@@ -0,0 +1,373 @@
+ * Users can always withdraw their tokens via multi-asset withdraws.
+ *
+ * MetaSwap is a modified version of Swap that allows Swap's LP token to be utilized in pooling with other tokens.
+ *
+ * @dev Most of the logic is stored as a library `SwapUtils` for the sake of reducing contract's
+ * deployment size.
+ */
+contract MetaSwap is Swap {
+    using MetaSwapUtils for SwapUtils.Swap;
+
+    MetaSwapUtils.MetaSwap public metaSwapStorage;
+
+    uint256 constant MAX_UINT256 = 2**256 - 1;
+
+    /*** EVENTS ***/
+
+    // events replicated from SwapUtils to make the ABI easier for dumb
+    // clients
+    event TokenSwapUnderlying(
+        address indexed buyer,
+        uint256 tokensSold,
+        uint256 tokensBought,
+        uint128 soldId,
+        uint128 boughtId
+    );
+
+    /**
+     * @notice Initializes this Swap contract with the given parameters.
+     * This will also deploy the LPToken that represents users
+     * LP position. The owner of LPToken will be this contract - which means
+     * only this contract is allowed to mint new tokens.
+     *
+     * @param _pooledTokens an array of ERC20s this pool will accept
+     * @param decimals the decimals to use for each pooled token,
+     * eg 8 for WBTC. Cannot be larger than POOL_PRECISION_DECIMALS
+     * @param lpTokenName the long-form name of the token to be deployed
+     * @param lpTokenSymbol the short symbol for the token to be deployed
+     * @param _a the amplification coefficient * n * (n - 1). See the
+     * StableSwap paper for details
+     * @param _fee default swap fee to be initialized with
+     * @param _adminFee default adminFee to be initialized with
+     * @param _withdrawFee default withdrawFee to be initialized with
+     */
+    function initialize(
+        IERC20[] memory _pooledTokens,
+        uint8[] memory decimals,
+        string memory lpTokenName,
+        string memory lpTokenSymbol,
+        uint256 _a,
+        uint256 _fee,
+        uint256 _adminFee,
+        uint256 _withdrawFee,
+        ISwap baseSwap
+    ) public virtual initializer {
+        Swap.initialize(
+            _pooledTokens,
+            decimals,
+            lpTokenName,
+            lpTokenSymbol,
+            _a,
+            _fee,
+            _adminFee,
+            _withdrawFee
+        );
+
+        // MetaSwap initializer
+        metaSwapStorage.baseSwap = baseSwap;
+        metaSwapStorage.baseVirtualPrice = baseSwap.getVirtualPrice();
+        metaSwapStorage.baseCacheLastUpdated = block.timestamp;
+        for (uint8 i; i < 32; i++) {
+            try baseSwap.getToken(i) returns (IERC20 token) {
+                token.approve(address(baseSwap), MAX_UINT256);
+                metaSwapStorage.baseTokens.push(token);
+            } catch {
+                break;
+            }
+        }
+        _pooledTokens[_pooledTokens.length - 1].approve(
+            address(baseSwap),
+            MAX_UINT256
+        );
```

Comment:
can you add a comment here explaining what this is for? (e.g. does `_pooledTokens` have to include the new non-base ERC20 as the last item?)